client/dwh_migration_client/macro_expander.py

```diff
@@ -0,0 +1,273 @@
+            generated = full_match
+        if not file_name in self.reverse:
+            self.reverse[file_name] = {}
+        reverse_for_file = self.reverse[file_name]
+        if generated in reverse_for_file:
+            reverse_for_file[generated] = (
+                reverse_for_file[generated][0] + 1,
+                reverse_for_file[generated][1].union({full_match}),
+            )
+        else:
+            reverse_for_file[generated] = (1, {full_match})
+        return generated
+
+    def expand(self, file_name: str, text: str) -> str:
+        """
+        Expands all macros in the given input text
+
+        Args:
+            file_name: The name of the file that the 'text' comes from
+            text: The text containing macro uses
+        Returns: 'text' with its macro uses expanded
+        """
+        return self.pattern.sub(
+            lambda match: self._substitution(file_name, match), text
+        )
+
+    def un_expand(self, file_name: str, text: str) -> str:
+        """
+        Reverses the macro expansion done by 'expand'.
+
+        Args:
+            file_name: The name of the file that the 'text' comes from.
+            text: The text resulting from invoking this expander on some input.
+        Returns: 'text' with the macro expansions undone.
+        """
+        if not file_name in self.reverse:
+            return text
+        self._sanity_check(file_name)
+        for replacement, original in self.reverse[file_name].items():
+            result = None
+            original_text = original[1].pop()
+            if self.un_generator:
+                result = re.subn(
+                    re.escape(replacement),
+                    self.un_generator(file_name, replacement, original_text),
+                    text,
+                    flags=re.I,
+                )
+            else:
+                result = re.subn(re.escape(replacement), original_text, text)
+            text = result[0]
+        return text
+
+
+class MacroExpanderRouter(RecordingLogger):
+    """
+    Contains multiple 'MacroExpander' instances and decided which to use
+    based on file names and file glob patterns.
+    """
+
+    all_macros: Dict[str, MacroExpander]
+
+    def __init__(self, all_macros: Dict[str, MacroExpander]) -> None:
+        """
+        Args:
+            all_macros: A mapping of file glob patterns to macro expanders.
+        """
+        super().__init__()
+        self.all_macros = all_macros
+        self.messages = []
+
+    def _choose_expander(self, file_name: str) -> Optional[MacroExpander]:
+        chosen_expander = None
+        chosen_pattern = None
+        matches = []
+        for pattern, expander in self.all_macros.items():
+            if fnmatch(file_name, pattern):
+                matches.append(file_name)
+                chosen_expander = expander
+                chosen_pattern = pattern
+        if len(matches) == 0:
+            return None
+        if len(matches) > 1:
```

Comment:
I am not sure if we want to handle this case this way.
I feel it makes sense that one file name matches multiple patterns, as long as the aggregated macros are different.  WDYT?

Reply:
What do you mean by "as long as the aggregated macros are different"? This is the code that decides which macro expander to use. Are you asking for the code to use all the matching expanders and then only keep the result of the expander that did the most macro expansions?

Reply:
Using the following definition as an example:

'*.sql':
    '{COMMON_MACRO}': 'replacement_1'
'special.sql':
    '{SPECIAL_MACRO}': 'replacement_2'

If I read your codes correctly, for the file "special.sql", only one macro will be substituted in this file.  I feel it makes sense that both macros should work here.  What do you think?

Reply:
That seems reasonable.